risc0/bigint2/src/ec/mod.rs

```diff
@@ -160,31 +179,71 @@ impl<const WIDTH: usize, C: Curve<WIDTH>> AffinePoint<WIDTH, C> {
     }
 
+    /// Elliptic curve doubling of the affine point.
     #[stability::unstable]
     pub fn double(&self, result: &mut Self) {
         let curve = C::CURVE;
-        double_raw(self.as_u32s(), curve.as_u32s(), &mut result.buffer);
+        // If the point is zero, can short-circuit and return the identity point.
+        if self.is_zero {
+            *result = *self;
+        } else {
+            if self.buffer[1] != [0u32; WIDTH] {
+                unsafe {
+                    double_raw(self.as_u32s(), curve.as_u32s(), &mut result.buffer);
+                }
+                // DO NOT REMOVE: the result is unchecked, and only the buffer is updated above
+                result.is_zero = false;
+            } else {
+                // DO NOT REMOVE: in this case a zero has been computed and the buffer is ignored
+                result.is_zero = true;
+            }
+        }
     }
 
+    /// Elliptic curve addition of the affine point.
     #[stability::unstable]
     pub fn add(&self, rhs: &AffinePoint<WIDTH, C>, result: &mut AffinePoint<WIDTH, C>) {
         let curve = C::CURVE;
-        // TODO: Do we want to check for P + P, P - P? It isn't necessary for soundness -- it will fail
-        // an EQZ if you try -- but maybe a pretty error here would be good DevEx?
-        add_raw(
-            self.as_u32s(),
-            rhs.as_u32s(),
-            curve.as_u32s(),
-            &mut result.buffer,
-        );
+
+        // If the left or right value is zero, can return the other value.
+        if self.is_zero {
+            *result = *rhs;
+        } else if rhs.is_zero {
+            *result = *self;
+        } else if self.buffer[0] == rhs.buffer[0] {
+            // X coordinates are the same, so either we double the value if it's the same point,
+            // or return the identity if it's different (not on the curve).
+            if self.buffer[1] != rhs.buffer[1] {
+                // x == x, y == -y, so the result is the identity point
+                result.is_zero = true;
+            } else {
+                // P + P, which can be done with a double call.
+                unsafe {
+                    double_raw(self.as_u32s(), curve.as_u32s(), &mut result.buffer);
+                }
+                // DO NOT REMOVE: the result is unchecked, and only the buffer is updated above
+                result.is_zero = false;
+            }
+        } else {
+            // X coordinates are different, so we can add the points as normal.
+            unsafe {
+                add_raw(
+                    self.as_u32s(),
+                    rhs.as_u32s(),
+                    curve.as_u32s(),
+                    &mut result.buffer,
+                );
+            }
+            // DO NOT REMOVE: the result is unchecked, and only the buffer is updated above
+            result.is_zero = false;
+        }
     }
 }
 
-fn double_raw<const WIDTH: usize>(
+unsafe fn double_raw<const WIDTH: usize>(
```

Comment:
Why is this unsafe?

Reply:
Just because the zero point handling isn't done for the raw functions and can panic in the host. Depends what the semantics of unsafe we want to follow, I just assumed that a panic in the host instead of guest would qualify.

Happy to switch back —- but was more of a reason to ensure that in future if refactors, they are done with care, because it can very easily break edge cases without noticing.

I'll add some tests for the edge cases regardless.

Reply:
I think `unsafe` has a specific definition around undefined behavior in rust. Extending this definition I think can cause confusion, especially over the long term.

Reply:
I'll switch back, missed this comment